### PR TITLE
Fix signed comparison bug in range_check_cast

### PR DESCRIPTION
**Pull request by seelabs**

### CODE REVIEW

**Comment by seelabs** [comment]
Rebased onto b3

File: src/ripple/protocol/impl/STParsedJSON.cpp [hunk 1]
```diff
@@ -42,14 +42,24 @@ namespace ripple {
...
-static T range_check_cast (U value, T minimum, T maximum)
+template <typename U, typename S>
+constexpr
+std::enable_if_t<std::is_unsigned<U>::value && std::is_signed<S>::value, U>
+to_unsigned (S value)
 {
-    if ((value < minimum) || (value > maximum))
-        Throw<std::runtime_error> ("Value out of range");
+    if (value < 0 || std::numeric_limits<U>::max () < value)
+        throw std::runtime_error ("Value out of range");
```
Comment (HowardHinnant) [inline src/ripple/protocol/impl/STParsedJSON.cpp]: Should we be using our `Throw` facility instead?

Comment (seelabs) [inline src/ripple/protocol/impl/STParsedJSON.cpp]: Yes. Fixed.

**Comment by HowardHinnant** [comment]
👍  I think even VS likes the `constexpr`! (I wasn't sure about that). :-)

**Comment by seelabs** [comment]
Squashed and rebased onto 0.33.0-b4

File: src/ripple/protocol/impl/STParsedJSON.cpp [hunk 2]
```diff
@@ -192,8 +202,7 @@ static boost::optional<detail::STVar> parseLeaf (
                 }
 
                 ret = detail::make_stvar <STUInt8> (field,
```
Comment (nbougalis) [inline src/ripple/protocol/impl/STParsedJSON.cpp]: We've already checked the range above. We can replace this with a straight `static_cast<unsigned char>` without any problems.

File: src/ripple/protocol/impl/STParsedJSON.cpp [hunk 3]
```diff
@@ -204,8 +213,7 @@ static boost::optional<detail::STVar> parseLeaf (
...
```
Comment (nbougalis) [inline src/ripple/protocol/impl/STParsedJSON.cpp]: Same here: we've already checked the range above so a straight `static_cast<unsigned char>` will work.

**Comment by seelabs** [comment]
Pushed a commit to address @nbougalis observation that the range check is not needed in some places.

**Comment by nbougalis** [comment]
:+1:

**Comment by seelabs** [comment]
Squashed

**Comment by nbougalis** [comment]
Merged as b02ec47b4f1765336b436cb6d5c6d4117eb3d681

**Pull request closed without merging**